src/c++/perf_analyzer/infer_context.h

```diff
@@ -0,0 +1,364 @@
+  uint64_t seq_id_;
+  // The data stream id providing data for the sequence
+  uint64_t data_stream_id_;
+  // The number of queries remaining to complete the sequence
+  size_t remaining_queries_;
+  // A lock to protect sequence data
+  std::mutex mtx_;
+};
+
+// Holds the running status of the thread.
+struct ThreadStat {
+  ThreadStat() {}
+
+  // The status of the worker thread
+  cb::Error status_;
+  // The status of the callback thread for async requests
+  cb::Error cb_status_;
+  // TODO REFACTOR TMA-1046 -- This should be in the InferContext class
+  // The statistics of the InferContext
+  std::vector<cb::InferStat> contexts_stat_;
+  // A vector of request timestamps <start_time, end_time>
+  // Request latency will be end_time - start_time
+  TimestampVector request_timestamps_;
+  // A lock to protect thread data
+  std::mutex mu_;
+};
+
+/// The properties of an asynchronous request required in
+/// the callback to effectively interpret the response.
+struct AsyncRequestProperties {
+  AsyncRequestProperties() : sequence_end_(false), delayed_(true) {}
+  // The timestamp of when the request was started.
+  std::chrono::time_point<std::chrono::system_clock> start_time_;
+  // Whether or not the request is at the end of a sequence.
+  bool sequence_end_;
+  // Whether or not the request is delayed as per schedule.
+  bool delayed_;
+};
+
+/// Sends inference requests to the server
+class InferContext {
+ public:
+  InferContext(
+      const uint32_t id, const bool async, const bool streaming,
+      const bool on_sequence_model, const bool using_json_data,
+      const int32_t batch_size, const cb::BackendKind backend_kind,
+      const SharedMemoryType shared_memory_type,
+      std::unordered_map<std::string, SharedMemoryData>& shared_memory_regions,
+      const uint64_t start_sequence_id, const uint64_t sequence_id_range,
+      const size_t sequence_length, std::atomic<uint64_t>& curr_seq_id,
+      std::uniform_int_distribution<uint64_t>& distribution,
+      std::shared_ptr<ThreadStat> thread_stat,
+      std::vector<std::shared_ptr<SequenceStat>>& sequence_stat,
+      std::shared_ptr<DataLoader> data_loader,
+      std::shared_ptr<ModelParser> parser,
+      std::shared_ptr<cb::ClientBackendFactory> factory)
+      : id_(id), async_(async), streaming_(streaming),
+        on_sequence_model_(on_sequence_model),
+        using_json_data_(using_json_data), batch_size_(batch_size),
+        backend_kind_(backend_kind), shared_memory_type_(shared_memory_type),
+        shared_memory_regions_(shared_memory_regions),
+        start_sequence_id_(start_sequence_id),
+        sequence_id_range_(sequence_id_range),
+        sequence_length_(sequence_length), curr_seq_id_(curr_seq_id),
+        distribution_(distribution), thread_stat_(thread_stat),
+        sequence_stat_(sequence_stat), data_loader_(data_loader),
+        parser_(parser), factory_(factory), data_step_id_(id)
+  {
```

Comment:
Small nit but can we just put all implementations in the implementation file if we're gonna use it at all?

Reply:
Discussed offline, this can be done later, more of a repo-wide stylistic change.